completed the app

shubh299 opened the pull request.

shubh299 committed:
completed the app

shubh299 committed:
Added changes for complete app

shubh299 committed:
Added player as argument in game initialization instead of game creating objects

gekco commented on Tic-Tac-Toe-Python/TicTacToe.py:
```diff
@@ -0,0 +1,140 @@
+
+class TicTacToeGame:
+    def __init__(self, player1: TicTacToePlayer, player2: TicTacToePlayer, number_of_rows: Optional[int] = 3):
+        self._NUMBER_OF_ROWS: int = number_of_rows
```
why all capital for _NUMBER_OF_ROWS ? is this supposed to be a constant ?

shubh299 replied:
I think number of rows should remain constant for an object of TicTacToe.

gekco commented on Tic-Tac-Toe-Python/TicTacToe.py:
```diff
@@ -0,0 +1,140 @@
+            # checking ith column
+            ith_column = [self._grid[j][i] for j in range(self._NUMBER_OF_ROWS)]
+
+            if ith_column.count(self._player_1.player_id) == self._NUMBER_OF_COLUMNS:
```
i see a lot of "ith_column.count(self._player_1.player_id) == self._NUMBER_OF_COLUMNS" , this could be a common function probably saying are_all_elements_equal ?

gekco replied:
also is there a better way than doing count of an element in the list ?

shubh299 replied:
Common function can certainly limit the number of lines. Another way to check if all elements are same can be to convert list to set and check the number of elements in the set. Is it a good way or is there any other way?

gekco commented on Tic-Tac-Toe-Python/TicTacToe.py:
```diff
@@ -0,0 +1,140 @@
+class TicTacToeGame:
+    def __init__(self, player1: TicTacToePlayer, player2: TicTacToePlayer, number_of_rows: Optional[int] = 3):
+        self._NUMBER_OF_ROWS: int = number_of_rows
+        self._NUMBER_OF_COLUMNS: int = number_of_rows
```
is it possible to have number of columns !== number of rows ?

shubh299 replied:
Initially I thought may be in some variation this could be possible but certainly keeping 2 variables can create inconsistency. Will remove extra variable.

gekco commented on Tic-Tac-Toe-Python/Player.py:
```diff
@@ -0,0 +1,18 @@
+from abc import ABC, abstractmethod
+
+
+class TicTacToePlayer(ABC):
```
why did we make this as an abstract class ? this is currently only being used in TicTacToeUserPlayer

shubh299 replied:
There can be a feature to create a computer player in which the next_move function will be implemented differently and it will inherit TicTacToePlayer class.

gekco commented on Tic-Tac-Toe-Python/TicTacToe.py:
```diff
@@ -0,0 +1,140 @@
+    def _check_game_over(self, move_count: int, total_moves: int) -> bool:
+        return move_count < total_moves
+
+    def play(self) -> None:
```
this lets you play multiple games. A good thing here would be to have a method for a single game which is being used by a multiple game function . Also a multiple game may want to use a different TicTacToeGame Altogether. for example first game may be 3x3 but next might be 5x5 which will have you chnage the game object altogether.

```
@classmethod
def new(rows, cols):
     return TicTacToeGame(rows, cols)

def play(self):
    pass
```

and then in main

you can do 
```
while new_game:
      game = TicTacToeGame.new()
      game.play()
```

shubh299 replied:
Is following explanation correct?
The advantage of using this `new()` method over calling constructor is that we can interpret the input/arguments and return object based on that. This helps in keeping the constructor as simple as possible.

gekco commented on Tic-Tac-Toe-Python/main.py:
```diff
@@ -0,0 +1,9 @@
+player_2 = TicTacToeUserPlayer('O', 'O')
+game1 = TicTacToeGame(player_1,player_2)
+
+game1.play()
```
the logic for multiple game should be here

shubh299 replied:
This along with the previous comment means 
* shifting the logic for new game from TicTacToe.play() to main 
* creating a new object for every iteration of the loop 
* giving the option to user to specify the dimensions of the board

Is this correct?

shubh299 committed:
Modifications to TicTacToe, main and utility_functions

Modified in TicTacToe.py: play to handle only one game and _check_win to move redundant code to utility_functions.py
Modified in main.py: Handle multiple games here with option to change size of game.

shubh299 committed:
Moved test scripts to test folder

shubh299 committed:
Updated comments for methods in TicTacToe.py

shubh299 committed:
Modified _check_game_over and play method in TicTacToe.py

Modified the methods to check for game over from grid itself instead of taking an extra variable.

shubh299 committed:
Modified play method to resolve issue with check for new game

added not to continue the game if it is not over

shubh299 committed:
Added tests for TicTacToe.py

shubh299 committed:
Resolved bug in play() method

Add check for game over to prevent 'game over message' from always printing

shubh299 committed:
Changed TicTacToe.py

Converted _NUMBER_OF_ROWS to snake case and created different function for initializing grid to improve readability.
